src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,295 @@
+  int maxDist)
+            : Operation(qec) {
+  this->triple = triple;
+  this->maxDist = maxDist;
+  if (childLeft) {
+    this->childLeft = childLeft.value();
+  }
+  if (childRight) {
+    this->childRight = childRight.value();
+  }
+  
+  if (triple._s.isVariable() && triple._o.isVariable()) {
+    leftChildVariable = triple._s.getVariable();
+    rightChildVariable = triple._o.getVariable();
+  } else {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+}
+
+std::shared_ptr<SpatialJoin> SpatialJoin::addChild(
+      std::shared_ptr<QueryExecutionTree> child, Variable varOfChild){
+  if (varOfChild == leftChildVariable) {
+    childLeft = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else if (varOfChild == rightChildVariable) {
+    childRight = child;
+    return std::make_shared<SpatialJoin>(getExecutionContext(), triple.value(),
+          childLeft, childRight, maxDist);
+  } else {
+    AD_THROW("variable does not match");
+  }
+}
+
+bool SpatialJoin::isConstructed() {
+  if (childLeft && childRight) {
+    return true;
+  }
+  return false;
+}
+
+std::vector<QueryExecutionTree*> SpatialJoin::getChildren() {
+  if (!(childLeft || childRight)) {
+    AD_THROW("SpatialJoin needs two variables");
+  }
+
+  return {childLeft.get(), childRight.get()};
+}
+
+
+string SpatialJoin::getCacheKeyImpl() const {
+  if (childLeft && childRight) {
+    std::ostringstream os;
+    os << "SpatialJoin\nChild1:\n" << childLeft->getCacheKey() << "\n";
+    os << "Child2:\n" << childRight->getCacheKey() << "\n";
+    return std::move(os).str();
+  } else {
+    return "incomplete SpatialJoin class";
+  }
+}
+
+string SpatialJoin::getDescriptor() const {
+  return "Descriptor of SpatialJoin";
+}
+
+size_t SpatialJoin::getResultWidth() const {
+  if (childLeft && childRight) {
+    // don't subtract anything because of a common join column. In the case
+    // of the spatial join, the join column is different for both sides (e.g.
+    // objects with a distance of at most 500m to each other, then each object
+    // might contain different positions, which should be kept)
+    if (addDistToResult) {
+      return childLeft->getResultWidth() + childRight->getResultWidth() + 1;
+    } else {
+      return childLeft->getResultWidth() + childRight->getResultWidth();
+    }
+  } else {
+    return 1;  // dummy return, as the class does not have its children yet
+  }
+}
+
+size_t SpatialJoin::getCostEstimate() {
+  return 50000;  // dummy return for now
+}
+
+uint64_t SpatialJoin::getSizeEstimateBeforeLimit() {
+  return 1;  // dummy return for now
+}
+
+float SpatialJoin::getMultiplicity(size_t col) {
+  return 1;  // dummy return for now
+}
+
+bool SpatialJoin::knownEmptyResult() {
+  return false;  // dummy return for now
+}
+
+vector<ColumnIndex> SpatialJoin::resultSortedOn() const {
+  return {};  // dummy return for now
```

Comment:
For the current implementation you probably preserve something, depending on the way your loop is set up, check it and report it.